### PR TITLE
Prevent Early Ordering Pushdown to Enable Aggregation Pushdown to MySQL

### PR DESCRIPTION
**Pull request by systay**
## Description
This PR introduces a crucial check in the `pushOrderingUnderAgg`r method to prevent the premature pushdown of the `ORDER BY` clause during the query planning process. By ensuring that the ordering is not pushed down too early, we allow the aggregation to be effectively pushed down to MySQL, optimizing query execution.

### Added a Check for Planning Phase:
* When attempting to push Ordering under Aggregation, we first check which phase we are in.
* This check ensures that the ordering is only pushed down after reaching the appropriate phase in the query planning process.
* If the planning phase has not yet reached delegateAggregation, the method returns early, leaving the ordering intact.
Rationale:

### Optimization:
* Pushing down the ORDER BY clause too early can interfere with the ability to push down the aggregation to MySQL.
* Aggregation pushdown to MySQL is crucial for optimizing query performance, as it allows MySQL to handle the aggregation, reducing the data that needs to be processed by the application.
* By preventing early ordering pushdown, we maintain the potential for aggregation pushdown, resulting in more efficient query execution.

### Impact:
* This change primarily affects the query planning phase, ensuring that the ORDER BY clause is only pushed down at the correct time.
* It improves the overall performance of queries involving aggregation by leveraging MySQL's capabilities more effectively.

## Related Issue(s)
Issue: #16279

## Checklist
-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/operators/aggregation_pushing.go [hunk 1]
```diff
@@ -369,14 +369,14 @@ func pushAggregationThroughApplyJoin(ctx *plancontext.PlanningContext, rootAggr
 
 	columns := &applyJoinColumns{}
 	output, err := splitAggrColumnsToLeftAndRight(ctx, rootAggr, join, !join.JoinType.IsInner(), columns, lhs, rhs)
-	join.JoinColumns = columns
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operators/aggregation_pushing.go:372]: This bug was not related to the rest of the PR, but it caused tests to fail, so it had to go.

File: go/vt/vtgate/planbuilder/operators/apply_join.go [hunk 2]
```diff
@@ -224,9 +216,12 @@ func (aj *ApplyJoin) getJoinColumnFor(ctx *plancontext.PlanningContext, orig *sq
 	case deps.IsSolvedBy(both):
 		col = breakExpressionInLHSandRHS(ctx, e, TableID(aj.LHS))
 	default:
-		panic(vterrors.VT13002(sqlparser.String(e)))
+		panic(vterrors.VT13001(fmt.Sprintf("expression depends on tables outside this join: %s", sqlparser.String(e))))
 	}
 
+	col.GroupBy = addToGroupBy
```
Comment (systay) [inline go/vt/vtgate/planbuilder/operators/apply_join.go:222]: This is another bug that was not related to the issue being found, but it caused some plans to be missing group by expressions

**Review by GuptaManan100 — APPROVED** [review]
LGTM

**Review by harshit-gangal — COMMENTED** [review]
could you publish some benchmarks as the PR description claims the impact is of improved performance?

File: go/vt/vtgate/planbuilder/testdata/postprocess_cases.json [hunk 3]
```diff
@@ -2088,16 +2088,16 @@
           {
             "OperatorType": "Aggregate",
             "Variant": "Ordered",
-            "Aggregates": "min(1|3) AS min(a.id)",
-            "GroupBy": "(0|2)",
-            "ResultColumns": 2,
+            "Aggregates": "any_value(2)",
+            "GroupBy": "(0|1)",
+            "ResultColumns": 3,
             "Inputs": [
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/testdata/postprocess_cases.json:2094]: Filter above will receive random value of `min(a.id)` this looks unexpected, we should perform the min aggregation of min coming from the route below and then send that to the filter above.

**Review by harshit-gangal — COMMENTED** [review]
We should add some e2e tests for the fix and any new cases added in the plan tests.

File: go/vt/vtgate/planbuilder/testdata/tpch_cases.json
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/testdata/tpch_cases.json:1]: these are pretty long plans to validate. can you run the vitess-tester for it and also we should enabled benchmarking for it in arewefastyet

**Review by harshit-gangal — COMMENTED** [review]
I could not find how we improved the planning unless the improvements are in tpch cases.

**Comment by systay** [comment]
> I could not find how we improved the planning unless the improvements are in tpch cases.

I added a plan to the bottom of `aggr_cases.json`. 

https://github.com/vitessio/vitess/pull/16278/files#diff-991c94d18be04f57568bac5c003b73009599842b38ee859c489fa286325b9b30R7150

That query produces a very different plan on `main`:

```json
{
  "QueryType": "SELECT",
  "Original": "select sum(user.type) from user join user_extra on user.team_id = user_extra.id group by user_extra.col order by user_extra.col",
  "Instructions": {
    "OperatorType": "Aggregate",
    "Variant": "Ordered",
    "Aggregates": "sum(0) AS sum(`user`.type)",
    "GroupBy": "1",
    "ResultColumns": 1,
    "Inputs": [
      {
        "OperatorType": "Sort",
        "Variant": "Memory",
        "OrderBy": "1 ASC",
        "Inputs": [
          {
            "OperatorType": "Join",
            "Variant": "Join",
            "JoinColumnIndexes": "L:0,R:0",
            "JoinVars": {
              "user_team_id": 1
            },
            "TableName": "`user`_user_extra",
            "Inputs": [
              {
                "OperatorType": "Route",
                "Variant": "Scatter",
                "Keyspace": {
                  "Name": "user",
                  "Sharded": true
                },
                "FieldQuery": "select `user`.type, `user`.team_id from `user` where 1 != 1",
                "Query": "select `user`.type, `user`.team_id from `user`",
                "Table": "`user`"
              },
              {
                "OperatorType": "Route",
                "Variant": "Scatter",
                "Keyspace": {
                  "Name": "user",
                  "Sharded": true
                },
                "FieldQuery": "select user_extra.col from user_extra where 1 != 1",
                "Query": "select user_extra.col from user_extra where user_extra.id = :user_team_id",
                "Table": "user_extra"
              }
            ]
          }
        ]
      }
    ]
  },
  "TablesUsed": [
    "user.user",
    "user.user_extra"
  ]
}
```

As you can see here, MySQL is not doing any aggregation at all.

File: go/vt/vtgate/planbuilder/testdata/tpch_cases.json
Comment (systay) [reply go/vt/vtgate/planbuilder/testdata/tpch_cases.json:1]: the tpch queries are tested in [end-to-end tests](https://github.com/vitessio/vitess/blob/main/go/test/endtoend/vtgate/vitess_tester/tpch/tpch.test) already

**Comment by systay** [comment]
```
                                                 │    old.txt    │               new.txt                │
                                                 │    sec/op     │    sec/op     vs base                │
ShardedAggrPushDown/user-100-user_extra-100-10      15.41m ±  6%   16.71m ± 12%   +8.41% (p=0.029 n=10)
ShardedAggrPushDown/user-100-user_extra-500-10      15.21m ±  3%   15.52m ±  1%   +1.99% (p=0.015 n=10)
ShardedAggrPushDown/user-100-user_extra-1000-10     15.23m ±  2%   15.59m ±  1%   +2.38% (p=0.005 n=10)
ShardedAggrPushDown/user-500-user_extra-100-10      73.57m ±  1%   46.18m ±  3%  -37.23% (p=0.000 n=10)
ShardedAggrPushDown/user-500-user_extra-500-10      74.65m ±  1%   76.05m ±  1%   +1.87% (p=0.002 n=10)
ShardedAggrPushDown/user-500-user_extra-1000-10     77.46m ± 12%   76.23m ±  2%        ~ (p=0.796 n=10)
ShardedAggrPushDown/user-1000-user_extra-100-10    143.79m ±  2%   58.23m ±  1%  -59.51% (p=0.000 n=10)
ShardedAggrPushDown/user-1000-user_extra-500-10     146.1m ±  1%   131.8m ±  1%   -9.78% (p=0.000 n=10)
ShardedAggrPushDown/user-1000-user_extra-1000-10    146.3m ±  2%   151.8m ±  1%   +3.80% (p=0.000 n=10)
geomean                                             55.08m         47.63m        -13.53%

                                                 │    old.txt    │                new.txt                │
                                                 │     B/op      │     B/op       vs base                │
ShardedAggrPushDown/user-100-user_extra-100-10     11.39Ki ±  1%   11.46Ki ±  2%        ~ (p=0.796 n=10)
ShardedAggrPushDown/user-100-user_extra-500-10     11.48Ki ±  4%   11.46Ki ±  4%        ~ (p=0.780 n=10)
ShardedAggrPushDown/user-100-user_extra-1000-10    11.39Ki ±  1%   11.37Ki ±  1%        ~ (p=0.837 n=10)
ShardedAggrPushDown/user-500-user_extra-100-10     13.34Ki ±  8%   12.09Ki ±  8%   -9.40% (p=0.019 n=10)
ShardedAggrPushDown/user-500-user_extra-500-10     49.11Ki ±  2%   49.13Ki ±  2%        ~ (p=0.962 n=10)
ShardedAggrPushDown/user-500-user_extra-1000-10    50.10Ki ±  2%   49.62Ki ±  1%        ~ (p=0.267 n=10)
ShardedAggrPushDown/user-1000-user_extra-100-10    15.34Ki ± 13%   12.46Ki ± 10%  -18.79% (p=0.001 n=10)
ShardedAggrPushDown/user-1000-user_extra-500-10    50.43Ki ±  5%   50.11Ki ±  4%   -0.65% (p=0.048 n=10)
ShardedAggrPushDown/user-1000-user_extra-1000-10   93.17Ki ±  2%   93.16Ki ±  2%        ~ (p=0.122 n=10)
geomean                                            24.78Ki         23.92Ki         -3.48%

                                                 │   old.txt   │              new.txt               │
                                                 │  allocs/op  │  allocs/op   vs base               │
ShardedAggrPushDown/user-100-user_extra-100-10      220.0 ± 0%    216.0 ± 0%  -1.82% (p=0.000 n=10)
ShardedAggrPushDown/user-100-user_extra-500-10      220.0 ± 0%    216.0 ± 0%  -1.82% (p=0.000 n=10)
ShardedAggrPushDown/user-100-user_extra-1000-10     220.0 ± 0%    216.0 ± 0%  -1.82% (p=0.000 n=10)
ShardedAggrPushDown/user-500-user_extra-100-10      220.0 ± 0%    216.0 ± 0%  -1.82% (p=0.000 n=10)
ShardedAggrPushDown/user-500-user_extra-500-10     1.022k ± 0%   1.018k ± 0%  -0.39% (p=0.000 n=10)
ShardedAggrPushDown/user-500-user_extra-1000-10    1.022k ± 0%   1.018k ± 0%  -0.39% (p=0.000 n=10)
ShardedAggrPushDown/user-1000-user_extra-100-10     220.0 ± 0%    216.0 ± 0%  -1.82% (p=0.000 n=10)
ShardedAggrPushDown/user-1000-user_extra-500-10    1.022k ± 0%   1.018k ± 0%  -0.39% (p=0.000 n=10)
ShardedAggrPushDown/user-1000-user_extra-1000-10   2.023k ± 0%   2.019k ± 0%  -0.20% (p=0.000 n=10)
```

File: go/vt/vtgate/planbuilder/testdata/aggr_cases.json [hunk 4]
```diff
@@ -5324,8 +5324,8 @@
                           "Name": "user",
                           "Sharded": true
                         },
-                        "FieldQuery": "select count(*), :user_id + user_extra.id, weight_string(:user_id + user_extra.id) from user_extra where 1 != 1 group by :user_id + user_extra.id",
-                        "Query": "select count(*), :user_id + user_extra.id, weight_string(:user_id + user_extra.id) from user_extra group by :user_id + user_extra.id",
+                        "FieldQuery": "select count(*), :user_id + user_extra.id, weight_string(:user_id + user_extra.id) from user_extra where 1 != 1 group by :user_id + user_extra.id, weight_string(:user_id + user_extra.id)",
```
Comment (arthurschreiber) [inline go/vt/vtgate/planbuilder/testdata/aggr_cases.json:5327]: Does the additional grouping by weight string make sense? 🤔 If so, can you explain why?

File: go/test/endtoend/vtgate/queries/benchmark/benchmark_test.go [hunk 5]
```diff
@@ -5,7 +5,7 @@ Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
-    http://www.apache.org/licenses/LICENSE-2.0
+  http://www.apache.org/licenses/LICENSE-2.0
```
Comment (frouioui) [inline go/test/endtoend/vtgate/queries/benchmark/benchmark_test.go:8]: I don't think this is wanted

File: go/vt/vtgate/planbuilder/testdata/tpch_cases.json [hunk 6]
```diff
@@ -566,149 +566,208 @@
...
-                            "Values": [
-                              ":l_orderkey"
-                            ],
-                            "Vindex": "hash"
+                            "TableName": "lineitem_orders_supplier_nation",
+                            "Inputs": [
+                              {
+                                "OperatorType": "Projection",
+                                "Expressions": [
+                                  "sum(volume) * count(*) as revenue",
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/testdata/tpch_cases.json:618]: can you explain why we are not passing `volume` as an index, like we do for the other columns?

Comment (systay) [reply go/vt/vtgate/planbuilder/testdata/tpch_cases.json:618]: we are, the printing of the plan is just not showing it

File: go/vt/vtgate/planbuilder/testdata/aggr_cases.json (hunk 4 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/testdata/aggr_cases.json]: no, they don't really make sense

File: go/vt/vtgate/planbuilder/testdata/postprocess_cases.json (hunk 3 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/testdata/postprocess_cases.json]: good catch!

File: go/vt/vtgate/planbuilder/operators/offset_planning.go [hunk 7]
```diff
@@ -96,62 +95,67 @@ func useOffsets(ctx *plancontext.PlanningContext, expr sqlparser.Expr, op Operat
...
+		}
+		inputs := op.Inputs()
+		if len(inputs) > 1 {
+			panic(vterrors.VT12001("unexpected multiple inputs"))
+		}
+		src := inputs[0]
+		_, isRoute := src.(*Route)
+		if isRoute {
+			panic(vterrors.VT12001("failed to find the aggregator"))
+		}
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/operators/offset_planning.go:112]: not sure if these errors should be `unsupported` or `internal` errors.

Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/operators/offset_planning.go:112]: do we have a test case that produces these errors?

**Review by harshit-gangal — APPROVED** [review]

**Pull request merged**